R/helper_patient_data_fix.R

```diff
@@ -172,27 +172,89 @@ parse_dates <- function(date) {
 
 #' @title Check value against list of allowed values.
 #'
+#' @description
+#' Invalid values are logged and depending on the replace invalid flag
+#' also replaced with an error value. When checking for the validity of values
+#' the case of the string is not considered nor are spaces and non-alphanumeric
+#' characters. If the value matches an entry in the allowed_values list
+#' that match is returned as the result.
+#'
 #' @param x character value to check.
 #' @param valid_values list of valid options.
-#' @param error_val value that is used to signal invalid values.
 #' @param id patient id.
+#' @param replace_invalid A flag controlling whether to also replace invalid .
+#' @param error_val value that is used to signal invalid values.
 #' @param col column name when used with mutate/across.
 #'
 #' @return Either x or "Other".
-check_allowed_values <- function(x, valid_values, error_val, id, col = "") {
+check_allowed_values <- function(x, valid_values, id, replace_invalid = TRUE, error_val = "Undefined", col = "") {
```

Comment:
the naming is somewhat off. If the function is named `check_` something I would not expect a return value or a boolean, but we are returning either the original value or `x`. Maybe we should call it `validate_x` but I don't find a good name. Anyway, it is not the allowed values that we check but the invalid ones, so maybe at least we could rename it to `handle_invalid_values`.

Reply:
I was also wondering about this. And the fact that the function is very complex and does many things does not make it easier. In this version the function might actually modify valid values as well to align them with the provided validity list. For instance if `Foo` is a valid value `foo` will be considered valid and will be mapped to `Foo`. While this adds additional complexity I though it might add nicer downstream guarantees wrt. the "valid" values that are expected based on configuration files. Previously `foo` would have been valid and unchanged.

I was also thinking to maybe go for something as `validate_`  but havn't found a solution yet with which I'm happy.